gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/GenerateWorkUnitsImpl.java

```diff
@@ -0,0 +1,154 @@
+      fs.mkdirs(workDirRoot);
+
+      List<WorkUnit> workUnits = generateWorkUnitsForJobState(jobState, eventSubmitterContext, closer);
+
+      JobStateUtils.writeWorkUnits(workUnits, workDirRoot, jobState, fs);
+      JobStateUtils.writeJobState(jobState, workDirRoot, fs);
+
+      return workUnits.size();
+    } catch (ReflectiveOperationException roe) {
+      String errMsg = "Unable to construct a source for generating workunits for job " + jobState.getJobId();
+      log.error(errMsg, roe);
+      throw ApplicationFailure.newNonRetryableFailureWithCause(errMsg, "Failure: new Source()", roe);
+    } catch (IOException ioe) {
+      String errMsg = "Failed to generate workunits for job " + jobState.getJobId();
+      log.error(errMsg, ioe);
+      throw ApplicationFailure.newFailureWithCause(errMsg, "Failure: generating/writing workunits", ioe);
+    } finally {
+      // TODO: implement Troubleshooter integration!
+    }
+  }
+
+  protected static List<WorkUnit> generateWorkUnitsForJobState(JobState jobState, EventSubmitterContext eventSubmitterContext, Closer closer)
+      throws ReflectiveOperationException {
+    Source<?, ?> source = JobStateUtils.createSource(jobState);
+    WorkUnitStream workUnitStream = source instanceof WorkUnitStreamSource
+        ? ((WorkUnitStreamSource) source).getWorkunitStream(jobState)
+        : new BasicWorkUnitStream.Builder(source.getWorkunits(jobState)).build();
+
+    // TODO: report (timer) metrics for workunits creation
+    if (workUnitStream == null || workUnitStream.getWorkUnits() == null) { // indicates a problem getting the WUs
+      String errMsg = "Failure in getting work units for job " + jobState.getJobId();
+      log.error(errMsg);
+      // TODO: decide whether a non-retryable failure is too severe... (in most circumstances, it's likely what we want)
+      throw ApplicationFailure.newNonRetryableFailure(errMsg, "Failure: Source.getWorkUnits()");
+    }
+
+    if (!workUnitStream.getWorkUnits().hasNext()) { // no work unit to run: entirely normal result (not a failure)
+      log.warn("No work units created for job " + jobState.getJobId());
+      return Lists.newArrayList();
+    }
+
+    // TODO: count total bytes for progress tracking!
+
+    boolean canCleanUp = canCleanStagingData(jobState);
+    DestinationDatasetHandlerService datasetHandlerService = closer.register(
+        new DestinationDatasetHandlerService(jobState, canCleanUp, eventSubmitterContext.create()));
+    WorkUnitStream handledWorkUnitStream = datasetHandlerService.executeHandlers(workUnitStream);
+
+    // initialize writer and converter(s)
+    // TODO: determine whether registration here is effective, or the lifecycle of this activity is too brief (as is likely!)
+    closer.register(WriterInitializerFactory.newInstace(jobState, handledWorkUnitStream)).initialize();
+    closer.register(ConverterInitializerFactory.newInstance(jobState, handledWorkUnitStream)).initialize();
```

Comment:
I think these would not be useful as the processWorkUnit activity should be handling the writer initializers and converter initializers

Reply:
I just wasn't sure, because these take a `WorkUnitStream` param, but the subsequent "work fulfillment" processing wouldn't have one of those around--only "work discovery" does.  could these be adding props to the WUs before they're serialized and persisted?

another data point is that the `ProcessWorkUnitImpl` activity seems to work just fine even though it doesn't reference these factories.

I haven't had a chance to look into this yet...  do you know for instance under what circumstances they're absolutely essential to successful operation?

Reply:
I took a look and it seems like they're very tied to Mysql related writers, other writers use `Noop` classes for intialization. For mysql they do some preparation work in mysql tables before writing.
I think it's sort of similar to the destDatasetHandlerService except they are configured to be tied closely to the writer implementation.

Reply:
So I guess for now it's fine to keep it as-is until there's some time to clean up old abstractions